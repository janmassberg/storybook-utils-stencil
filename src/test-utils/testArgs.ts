export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: (): void => {
        // Implementation
    },
    onTestEvent: (event: Event): void => {
        // Handle event
        console.log(event.type);
    },
    _slot: "Test children",
};

export const testArgsButton = {
    required: true,
    variant: "primary",
    onClick: (event: Event) => {
        // Handle click event
        console.log(event.target);
    },
    _slot: "Label",
};

export const testArgsComponentGenerator = {
    component: "my-component",
    args: testArgs,
};

export const testCopmonentSlot = `<my-child-component foo-bar="foo">
<my-nested-child-component bar-foo="bar"><span class="test-class">Label</span></my-nested-child-component>
</my-child-component>`;
