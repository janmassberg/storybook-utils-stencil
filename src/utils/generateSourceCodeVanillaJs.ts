import { formatHtml } from "./formatHtml";
import { filterEventHandlers, filterJsxProperties, ArgsType } from "./argUtils";
import { objectToString } from "./sourceCodeUtils/utils";

export const generateVanillaJsCode = (
    component: string,
    props: [string, any][],
    events: [string, any][]
): string => {
    if (props.length === 0 && events.length === 0) {
        return "";
    }
    const jsProps =
        props.length > 0
            ? props
                .map(
                    ([key, value]) =>
                        `element.${key} = ${objectToString(value)};`
                )
                .join("\n")
            : "";
    const jsEvents =
        events.length > 0
            ? events
                .map(
                    ([key, value]) =>
                        `element.addEventListener("${key}", ${value.toString()});`
                )
                .join("\n")
            : "";
    return `
    const element = document.querySelector("${component}");
    ${jsProps}
    ${jsEvents}
    `;
};

export const generateSourceCodeVanillaJs = (
    component: string,
    args: any
): string => {
    const eventHandlers = filterEventHandlers(args);
    const jsxProps = filterJsxProperties(args);
    return formatHtml(
        `<script>${generateVanillaJsCode(component, jsxProps, eventHandlers)}</script>`);
};
