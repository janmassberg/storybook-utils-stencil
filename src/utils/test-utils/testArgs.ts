export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: () => {},
    onTestEvent: () => {},
    _children: "Test children",
};
