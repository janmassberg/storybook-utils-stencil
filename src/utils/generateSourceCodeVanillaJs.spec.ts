import { generateSourceCodeVanillaJs } from "./generateSourceCodeVanillaJs";
import { testArgs } from "../test-utils/testArgs";

describe("generateSourceCodeVanillaJs", () => {
    it("should return an empty string if no props or event handlers", () => {
        expect(
            generateSourceCodeVanillaJs("my-component", {
                testString: "test string",
                testNumber: 12345,
                testTrue: true,
                testFalse: false,
                testNull: null,
                testUndefined: undefined,
            })
        ).toEqual("");
    });

    it("should generate a script tag with prop injection and event handlers", () => {
        expect(
            generateSourceCodeVanillaJs("my-component", testArgs)
        ).toMatchSnapshot();
    });
});
