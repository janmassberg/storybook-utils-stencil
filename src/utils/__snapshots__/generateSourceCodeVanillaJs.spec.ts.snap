// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateSourceCodeVanillaJs should generate a script tag with prop injection and event handlers 1`] = `
"<script>
  const element = document.querySelector(\\"my-component\\");
  element.testArray = [1, 2, 3];
  element.testObject = { foo: \\"bar\\" };
  element.testFunction = function () {
    // do nothing
  };
  element.addEventListener(\\"onTestEvent\\", function (event) {
    console.log(event.type);
  });
</script>
"
`;
